docs/source/notes/amp_examples.rst

```diff
@@ -0,0 +1,157 @@
+-------------------------------
+
+The gradients resulting from ``scaler.scale(loss).backward()`` are scaled.  ``scaler.unscale`` can be
+used to unscale an optimizer's owned gradients in-place prior to ``scaler.step``.
+
+Gradient clipping
+"""""""""""""""""
+
+Gradient clipping involves manipulating a set of gradients such that their global norm
+(see :func:`torch.nn.utils.clip_grad_norm_`) or maximum magnitude (see :func:`torch.nn.utils.clip_grad_value_`)
+is :math:`<=` some user-imposed threshold.  Calling ``scaler.unscale(optimizer)`` before clipping enables you
+to clip unscaled gradients as usual::
+
+    scaler = AmpScaler()
+    ...
+    for input, target in data:
+        optimizer.zero_grad()
+        output = model(input)
+        loss = loss_fn(output, target)
+        scaler.scale(loss).backward()
+
+        # Unscale the gradients owned by optimizer in-place
+        scaler.unscale(optimizer)
+
+        # Since the optimizer's owned gradients are unscaled, clip as usual:
+        torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm)
+
+        scaler.step(optimizer)
+        scaler.update()
+
+``scaler`` records that ``scaler.unscale(optimizer)`` has been called for this optimizer
+this iteration, so ``scaler.step(optimizer)`` knows not to redundantly unscale gradients before
+calling ``optimizer.step()``.
+
+.. warning::
+    If you (optionally) choose to unscale gradients prior to stepping, ``scaler.unscale(optimizer)``
+    should only be invoked once per optimizer per step, and only after all gradients for that optimizer's
+    owned parameters have been accumulated.
+
+Working with Scaled Gradients
+-----------------------------
+
+For some operations, you may need to work with scaled gradients in a setting where
+``scaler.unscale`` is unsuitable.
+
+Gradient penalty
+""""""""""""""""
+
+A gradient penalty loss term requires manipulating scaled gradients.
+The gradient penalty example below demonstrates
+
+* creating scaled out-of-place gradients with :func:`torch.autograd.grad`, and
+* correct interaction of gradient scaling with double-backward.
+
+Here's how that looks for a simple L2 penalty::
+
+    scaler = AmpScaler()
+    ...
+    for input, target in data:
+        optimizer.zero_grad()
+        output = model(input)
+        loss = loss_fn(output, target)
+
+        # We should scale outputs for the out-of-place backward pass
+        grad_params = torch.autograd.grad(scaler.scale(loss), model.parameters(), create_graph=True)
+
+        # In general, the penalty term may depend nonlinearly on grad_params, so to be safe,
+        # manually unscale them before computing the penalty.  The unscale should be out-of-place
+        # and autograd-exposed.  For these reasons, and because grad_params are not owned by any optimizer,
+        # calling scaler.unscale(optimizer) here is unsuitable, and we use ordinary ops instead:
+        grad_params = [p*(1./scaler.get_scale()) for p in grad_params]
```

Comment:
So this is where @mruberry's concerns about `.to()` method for a scaler are valid - you are taking care of multi-device operation when you are doing the scaling, but if user chooses to do it manually, there will be errors in case of device mismatch.

Reply:
I see this as garden-variety tensor manipulation.  People always have to ensure that ops they invoke use tensors that are on the same device.  If people have gradients on multiple devices, they know this, and can do something like
```
scale = scaler.get_scale()
grad_params = [p*1./scale.to(p.device) for p in grad_params]
```
I'm in favor of that manual control over adding e.g. an additional unscale utility function that people can use on free gradients.

In this particular case, the unscale differs qualitatively from `scaler.unscale(optimizer)` in several important ways:

- It is out-of-place.
- It is autograd-exposed.
- The gradients are not owned by any optimizer.
- The inf/nan checking is irrelevant at this point.

IMO all these factors point away from using an AmpScaler utility and towards using ordinary ops.  Do I want some hypothetical utility function that acts on free gradients to have knobs `inplace=True|False` and `autograd_exposed = True|False` and `inf_relevant=True|False`?  I think that would be even more confusing.

The one benefit of adding a utility function that acts on a list of free gradients is we would have a control point to do it with multi tensor apply in the future, but we can just as easily update the example to have it show an ordinary use of multi tensor apply instead of an AmpScaler function that wraps some multi-tensor apply.

Reply:
I could add a `device=...` kwarg to `get_scale` that would serve a copy of the scale on a particular device, but I still think that for this particular use case letting the user call `.to` on the scale tensor manually is most intuitive.

Reply:
I can certainly add a `.to` method on AmpScaler itself to move the scale tensor.

However, the `found_inf` tensors are hard-associated with devices because they each represent the inf/nan state of gradients found on a given device for a given optimizer.  `AmpScaler.to()` could move the scale, but not these, which may be confusing for the user.

Reply:
get_scale now returns a Python float as discussed in Slack.